server/main_server/src/main.cpp

```diff
@@ -1009,6 +1028,21 @@ Mandatory arguments to long options are mandatory for short options too.
         }
 
         log_server::info("{}: Agent Factory PID = [{}].", __func__, g_pid_af);
+
+        // Wait for the agent factory's signal that it has completed initialization. This effectively
+        // means the agent factory has opened its listening socket and is entering its main loop.
+        while (0 == g_agent_factory_initialized) {
+            std::this_thread::sleep_for(std::chrono::milliseconds{250});
+            log_server::debug("{}: Waiting for agent factory to complete initialization.", __func__);
+
+            // If this loop cannot make process for some reason, allow the admin to stop the server
+            // without needing SIGKILL.
+            if (g_terminate || g_terminate_graceful) {
+                log_server::info("{}: Received shutdown instruction. Ending wait loop for agent factory initialization.", __func__);
+                return false;
+            }
+        }
```

Comment:
Should this loop terminate after _N_ iterations or _M_ seconds?

We don't do that anywhere else. Perhaps I'm being paranoid and trust the computer will work as intended as long as the design is correct.

Reply:
computers are different speeds... so feels like M seconds is what an admin will have a better feel for.

we *do* use timeouts for network things... but right, not sure about within the server itself...

as long as the admin can still send `g_terminate || g_terminate_graceful`, maybe we don't need a robust timeout here ... yet.

Reply:
I somewhat share your concern but I think we can introduce a configuration to time out after a while if this ever becomes an issue.

I'm tempted to say the log message should be bumped up to `info`. If the concern is that we might enter an infinite loop, one might like to know why the server is stuck. There's not really a way to change the log level and see what's going on if it's hanging once the server has started. However, if it takes 1 second from the start of this loop to start the agent factory, you'll get 4 messages in the log at the default level every time the agent factory has to be restarted (which could get annoying).

Reply:
Given this kind of thing exists in other parts of the redesign, I think we just let it ride until someone reports it as an issue. The server has not shown any behavior that makes me think the loop is an issue.

Will think about the log level bump.

Reply:
I'm going to leave the log level at debug for now (partly for the reason that @alanking laid out). We can also adjust log levels, etc. My hope is that if an issue does appear at this level, admins can restart the server and reproduce it.

There may also be room for thinking about making the server auto-bump log levels after some criteria is met - e.g. once log message _L_ is executed _N_ times, change the log level of _L_ to a more severe level.

Reply:
Created #8453.